client/hijack.go

```diff
@@ -140,23 +167,37 @@ func tlsDialWithDialer(dialer *net.Dialer, network, addr string, config *tls.Con
 	return &tlsClientCon{conn, rawConn}, nil
 }
 
-func dial(proto, addr string, tlsConfig *tls.Config) (net.Conn, error) {
+func dial(ctx context.Context, proto, addr string, tlsConfig *tls.Config) (net.Conn, error) {
+	// 0 = no timeout
+	var timeout time.Duration
+
+	deadline, ok := ctx.Deadline()
```

Comment:
Can we leave this only for the `npipe` case and everything else can use the context directly?

Reply:
`tlsDialWithDialer` requires translating to a timeout, as well. `net.DialContext` is the 1 exception here, unfortunately. I'm happy to:

*  move context->timeout translation down to `tlsDialWithDialer` & the `npipe` block from here
*  change `tlsDialWithDialer` to `tlsDialWithContext`

sounds good?

Reply:
`tlsDialWithContext` is what I thought. If we keep using the dialer, possibly the correct solution would be to wrap it early with context cancellation calling `dialer.Cancel` instead of the timeout logic.

Generally, if the stdlib or helper packages do not provide good ways to implement some cases we should try to fix them in these packages and ignore these cases until then. Having these exceptions and potentially leaking goroutines is worse than a little delay on cancellation. For example, I don't think the extra goroutine in `Handshake` is justified. My suggestion would be to either use the context directly and ignore the cases that can't do `DialContext` or convert to dialer once and only use dialer for everything.

Reply:
I did some research on this, and pursuant to https://github.com/rdallman/moby/blob/ccd30e464ac89d8bb2b4b64fb7b57128c721dc52/client/hijack.go#L58-L62 I believe that we can use https://golang.org/pkg/crypto/tls/#Conn.CloseWrite now in stdlib, and toss out the custom implementation here to be able to close the write part of the connection. woo. I will try to go ahead with this, and move this code to a `client/hijack_go17.go` file for older go versions as seems the convention here https://github.com/rdallman/moby/blob/ccd30e464ac89d8bb2b4b64fb7b57128c721dc52/client/tlsconfig_clone_go17.go -- we will still convert our context deadline into a dial deadline, but this seems fine in the grand scheme of things.

the code itself does spawn a goroutine https://golang.org/src/crypto/tls/tls.go?s=3255:3347#L148 the same as we do for the handshake timeout.

this leaves us with only the hijack do goroutine explicitly in the client code. after some auditing, it appears to use `(http.Request).Write` and `http.ReadResponse` which will respect the context, and I will remove that addition, thanks for pointing this out.